### ISSUE TITLE
v5.2.0 Connection open issue

### ISSUE THREAD

**Issue by deaklasECB** [dotnet/SqlClient#2422]
### Describe the bug
When upgrading from 5.1.5 to 5.2.0 and connecting to Azure Sql, I get an error when connecting to the database.

```
Exception message:
A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 26 - Error Locating Server/Instance Specified) Success 

Stack trace:
Microsoft.Data.SqlClient.SqlException:
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.TdsParser.Connect (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds..ctor (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnectionFactory.CreateConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnection.Open (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.Data.SqlClient.SqlConnection.Open (Microsoft.Data.SqlClient, Version=5.0.0.0, Culture=neutral, PublicKeyToken=23ec7fc2d6eaa4a5)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerConnection.OpenDbConnection (Microsoft.EntityFrameworkCore.SqlServer, Version=8.0.3.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)

```

### Expected behavior
Connection should be established.

I suspect this might be related to the [redirect/proxy connection policy](https://learn.microsoft.com/en-us/azure/azure-sql/database/connectivity-architecture?view=azuresql) in Azure SQL, because I have 2 instances, one with such a policy and one without, and in one case version 5.2.0 works but in the other it does not.

Is there anything to be changed in the connection string or an env. var to enable the connection?

### Further technical details
Microsoft.Data.SqlClient version: 5.2.0
.NET target: (.NET 8)
SQL Server version: Azure SQL
Operating system: Linux containers

**Comment by Wraith2** [comment]
It's possible that this is related to an unintentional change reported here https://github.com/dotnet/SqlClient/issues/2415 Could you try flipping the appcontext switch mentioned in that issue and see if it helps?

**Comment by deaklasECB** [comment]
I have set:

```csharp
AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseOneSecFloorInTimeoutCalculationDuringLogin", true);
```

but I still get the same error as before.

**Comment by ladeak** [comment]
Does the new release use a port or protocol during connection that the previous did not use?

What is the best way to 'debug' this, should I create an event listener and subscribe to traces emitted?

**Comment by deaklasECB** [comment]
After further testing:
- Using Azure SQL Redirect mode from Linux (asp.net core 8 chiseled image) does not work.
- Using Azure SQL Proxy mode from Linux works.
- Using Azure SQL from Windows in Redirect mode or Proxy mode works.

The issue is with the 1st case.

**Comment by JRahnama** [comment]
Can you try adding a port number at the end of connection string? like: tcp:<ip address>, port number?

**Comment by deaklasECB** [comment]
It is in the connection string:
`Server=tcp:sql-....database.windows.net,1433;Initial Catalo...j;MultipleActiveResultSets=False;Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;`

**Comment by deaklasECB** [comment]
It is using username and password authentication (removed that section from the connectionstring in the above samples). Let me test the package (I might only come back with results on Monday because I am catching a flight very soon).

**Comment by deaklasECB** [comment]
In my initial test from my dev vm, it seems to have fixed the issue. The version number for me was shown as 6.0.0-pull.106802. 
I will do a test on Monday from the container image as well if you think it worth it.
Thank you very much for the quick turnaround.

**Comment by JRahnama** [comment]
> In my initial test from my dev vm, it seems to have fixed the issue. The version number for me was shown as 6.0.0-pull.106802. I will do a test on Monday from the container image as well if you think it worth it. Thank you very much for the quick turnaround.

It will help to confirm that the regression is addressed in this build. Thank you.

**Comment by deaklasECB** [comment]
I have tested from the docker container as well, and I can confirm the fix works as expected. Thank you very much for the quick turnaround.

May I ask what was the issue? (just for my curiosity).

My last question: do you already have any plans for the next release? When would you expect the fix to go live? - Asking only, so that I can plan along.

**Comment by JRahnama** [comment]
@deaklasECB Issue #2378 was the initial report highlighting a regression introduced in v5.2.0. This issue will be resolved in the upcoming hotfix release, v5.2.1, with the release date yet to be determined.

**Comment by DavoudEshtehari** [comment]
Closing as duplicate of #2378.
